File: backend/app/utils/dialogue_text.py
import re


def strip_speaker_prefix(text: str, speaker: str | None = None) -> str:
    stripped = text.strip().strip('"“”')
    if not speaker:
        return stripped

    return re.sub(
        rf'^\s*["“”]?\s*{re.escape(speaker.strip())}\s*[:：-]\s*',
        "",
        stripped,
        flags=re.IGNORECASE,
    ).strip().strip('"“”')


def strip_duplicate_dialogue_from_narration(
    narration: str,
    dialogue: str | None,
    speaker: str | None = None,
) -> str:
    narration_text = narration.strip()
    if not dialogue:
        return narration_text

    dialogue_core = strip_speaker_prefix(dialogue, speaker)
    candidates = {dialogue.strip(), dialogue_core}
    if speaker and dialogue_core:
        candidates.add(f"{speaker.strip()}: {dialogue_core}")

    for candidate in sorted((c for c in candidates if c), key=len, reverse=True):
        narration_text = re.sub(
            re.escape(candidate),
            "",
            narration_text,
            flags=re.IGNORECASE,
        )

    if dialogue_core:
        narration_text = re.sub(
            r'["“”]([^"“”]+)["“”]',
            lambda match: "" if dialogue_core.lower() in match.group(1).lower() else match.group(0),
            narration_text,
            flags=re.IGNORECASE,
        )

    attribution_names = []
    if speaker:
        attribution_names.append(re.escape(speaker.strip()))
    attribution_names.extend(["hắn", "y", "gã", "nàng", "cô", "anh", "lão"])
    attribution = "|".join(dict.fromkeys(name for name in attribution_names if name))
    if attribution:
        narration_text = re.sub(
            rf"(?:^|[\s,.;:!?])(?:{attribution})\s+"
            r"(?:noi|nói|hoi|hỏi|dap|đáp|quat|quát|thot len|thốt lên|gam giong|gằn giọng|lam bam|lẩm bẩm)"
            r"\s*[,.;:!?]*\s*$",
            "",
            narration_text,
            flags=re.IGNORECASE,
        )

    narration_text = re.sub(r"\s+([,.!?;:])", r"\1", narration_text)
    narration_text = re.sub(r"([,.!?;:]){2,}", r"\1", narration_text)
    narration_text = re.sub(r"\s+", " ", narration_text).strip(" ,.;:!?")
    return narration_text.strip()

File: web-app/src/features/render/lib/dialogueText.ts
const escapeRegExp = (value: string): string => value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");

export const stripSpeakerPrefix = (text: string, speaker?: string | null): string => {
  const stripped = text.trim().replace(/^["“”]+|["“”]+$/g, "");
  const normalizedSpeaker = speaker?.trim();
  if (!normalizedSpeaker) return stripped;

  return stripped
    .replace(new RegExp(`^\\s*["“”]?\\s*${escapeRegExp(normalizedSpeaker)}\\s*[:：-]\\s*`, "i"), "")
    .replace(/^["“”]+|["“”]+$/g, "")
    .trim();
};

export const stripDuplicateDialogueFromNarration = (
  narration: string,
  dialogue?: string | null,
  speaker?: string | null,
): string => {
  let narrationText = narration.trim();
  if (!dialogue?.trim()) return narrationText;

  const dialogueCore = stripSpeakerPrefix(dialogue, speaker);
  const candidates = new Set([dialogue.trim(), dialogueCore]);
  if (speaker?.trim() && dialogueCore) {
    candidates.add(`${speaker.trim()}: ${dialogueCore}`);
  }

  Array.from(candidates)
    .filter(Boolean)
    .sort((a, b) => b.length - a.length)
    .forEach((candidate) => {
      narrationText = narrationText.replace(new RegExp(escapeRegExp(candidate), "gi"), "");
    });

  if (dialogueCore) {
    narrationText = narrationText.replace(/["“”]([^"“”]+)["“”]/gi, (match, quoted: string) =>
      quoted.toLowerCase().includes(dialogueCore.toLowerCase()) ? "" : match,
    );
  }

  const attributionNames = [
    speaker?.trim(),
    "hắn",
    "y",
    "gã",
    "nàng",
    "cô",
    "anh",
    "lão",
  ].filter(Boolean) as string[];
  const attribution = Array.from(new Set(attributionNames.map(escapeRegExp))).join("|");
  if (attribution) {
    narrationText = narrationText.replace(
      new RegExp(
        `(?:^|[\\s,.;:!?])(?:${attribution})\\s+` +
          "(?:noi|nói|hoi|hỏi|dap|đáp|quat|quát|thot len|thốt lên|gam giong|gằn giọng|lam bam|lẩm bẩm)" +
          "\\s*[,.;:!?]*\\s*$",
        "i",
      ),
      "",
    );
  }

  return narrationText
    .replace(/\s+([,.!?;:])/g, "$1")
    .replace(/([,.!?;:]){2,}/g, "$1")
    .replace(/\s+/g, " ")
    .replace(/^[\s,.;:!?]+|[\s,.;:!?]+$/g, "")
    .trim();
};
